File: app/Store/src/features/profile/types/address.ts
// Context: Address type discriminator for shipping vs billing vs other
export type AddressType = 'Shipping' | 'Billing' | 'Other'

// Context: Full address entity returned by the address API
export interface Address {
  id: string
  userId: string
  addressType: AddressType
  firstName: string
  lastName: string | null
  address1: string
  address2: string | null
  city: string
  zipCode: string | null
  phone: string | null
  label: string | null
  isDefault: boolean
  countryName: string
  stateProvince: string | null
  countryCode: string | null
  stateCode: string | null
}

// Context: Address input payload for create and update operations
export interface AddressInput {
  addressType: AddressType
  firstName: string
  lastName?: string
  address1: string
  address2?: string
  city: string
  zipCode?: string
  phone?: string
  label?: string
  isDefault: boolean
  countryName: string
  stateProvince?: string
  countryCode?: string
  stateCode?: string
}

File: service/Api/src/Module/Customer/Domain/Addresses/Address.cs
using Module.Ordering.Domain.Orders;
using Module.Shipping.Domain.Shipments;

using Shared.Application.Domain.Models;

namespace Module.Customer.Domain.Addresses;

// Invariant: FirstName != null; Address1 != null; City != null; CountryName != null; IsDefault implies valid address type
public sealed partial class Address : Entity
{
    #region Properties

    public AddressType AddressType { get; set; }
    public string FirstName { get; set; } = string.Empty;
    public string? LastName { get; set; }
    public string Address1 { get; set; } = string.Empty;
    public string? Address2 { get; set; }
    public string City { get; set; } = string.Empty;
    public string? ZipCode { get; set; }
    public string? Phone { get; set; }
    public string? Label { get; set; }
    public bool IsDefault { get; set; }
    public bool IsDefaultBilling { get; set; }
    public bool IsDefaultShipping { get; set; }

    public string CountryName { get; set; } = string.Empty;
    public string? StateProvince { get; set; }
    public string? CountryCode { get; set; }
    public string? StateCode { get; set; }

    #endregion

    #region Relationships

    public Guid? UserProfileId { get; set; }
    public UserProfile? UserProfile { get; set; }

    public ICollection<Order> BillingOrders { get; set; } = [];
    public ICollection<Order> ShippingOrders { get; set; } = [];
    public ICollection<Shipment> Shipments { get; set; } = [];

    #endregion
}